pkg/logicalbackup/logicalbackup.go

```diff
@@ -194,42 +204,90 @@ func New(ctx context.Context, stopCh chan struct{}, cfg *config.Config) (*Logica
 		return nil, fmt.Errorf("could not prepare tables for backup: %v", err)
 	}
 
+	if err := lb.registerMetrics(); err != nil {
+		return nil, err
+	}
+
 	return lb, nil
 }
 
-func (b *LogicalBackup) saveTableMessage(tableOID dbutils.Oid, msg []byte) error {
+func (b *LogicalBackup) processDMLMessage(tableOID dbutils.Oid, cmd cmdType, msg []byte) error {
 	bt, ok := b.backupTables[tableOID]
 	if !ok {
 		log.Printf("table with OID %d is not tracked", tableOID)
 		return nil
 	}
 
+	// We need to write the transaction begin message to every table being collected,
+	// therefore, it cannot be written right away when we receive it. Instead, wait for
+	// the actual relation data to arrive and then write it here.
+	// TODO: avoid multiple writes
+
 	if _, ok := b.txBeginRelMsg[tableOID]; !ok {
-		ln, err := bt.SaveRawMessage(b.beginMsg, b.flushLSN)
-		if err != nil {
-			return fmt.Errorf("could not save begin message: %v", err)
+		if err := b.WriteCommandDataForTable(bt, msg, cBegin); err != nil {
+			return err
 		}
-
-		b.bytesWritten += ln
-		b.txBeginRelMsg[tableOID] = struct{}{}
+		b.txBeginRelMsg[bt.ID()] = struct{}{}
 	}
 
 	if b.typeMsg != nil {
-		ln, err := bt.SaveRawMessage(b.typeMsg, b.flushLSN)
-		if err != nil {
-			return fmt.Errorf("could not save type message: %v", err)
+		if err := b.WriteCommandDataForTable(bt, b.typeMsg, cType); err != nil {
+			return fmt.Errorf("could not write type message: %v", err)
 		}
-
-		b.bytesWritten += ln
 		b.typeMsg = nil
 	}
 
-	ln, err := bt.SaveRawMessage(msg, b.flushLSN)
+	if msg, ok := b.relationMessages[tableOID]; ok {
+		if err := b.WriteCommandDataForTable(bt, msg, cRelation); err != nil {
+			return fmt.Errorf("could not write a relation message: %v", err)
+		}
+		delete(b.relationMessages, tableOID)
+	}
+
+	if err := b.WriteCommandDataForTable(bt, msg, cmd); err != nil {
+		return fmt.Errorf("could not write message: %v", err)
+	}
+
+	return nil
+}
+
+func (b *LogicalBackup) WriteCommandDataForTable(t tablebackup.TableBackuper, msg []byte, cmd cmdType) error {
+	var promType string
+
+	ln, err := t.WriteDelta(msg, b.flushLSN)
 	if err != nil {
-		return fmt.Errorf("could not save message: %v", err)
+		return err
 	}
 
 	b.bytesWritten += ln
+	b.msgCnt[cmd]++
+
+	switch cmd {
```

Comment:
we can use map here instead of switch

Reply:
my concern is the introduced switch has no logic inside, so I would substitute it with the map

Reply:
A map is certainly less performant than a plain switch, and also takes more memory.

Reply:
I think the performance/memory difference is neglectable here.
my point is to have more clean code